pkg/operator/ceph/upgrade/upgrade.go

```diff
@@ -0,0 +1,171 @@
+)
+
+// CephDaemonsVersions is a structure that can be used to parsed the output of the 'ceph versions' command
+type CephDaemonsVersions struct {
+	Mon     map[string]int `json:"mon,omitempty"`
+	Mgr     map[string]int `json:"mgr,omitempty"`
+	Mds     map[string]int `json:"mds,omitempty"`
+	Overall map[string]int `json:"overall,omitempty"`
+}
+
+var (
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", "upgrade")
+)
+
+func getCephMonVersionString(context *clusterd.Context) (string, error) {
+	output, err := context.Executor.ExecuteCommandWithOutput(false, "", "ceph", "version")
+	if err != nil {
+		return "", fmt.Errorf("failed to run ceph tell: %+v", err)
+	}
+	logger.Debug(output)
+
+	return output, nil
+}
+
+func getCephMonVersionsString(context *clusterd.Context) (string, error) {
+	output, err := context.Executor.ExecuteCommandWithOutput(false, "", "ceph", "versions")
+	if err != nil {
+		return "", fmt.Errorf("failed to run ceph tell: %+v", err)
+	}
+	logger.Debug(output)
+
+	return output, nil
+}
+
+// GetCephMonVersion reports the Ceph version of all the monitors, or at least a majority with quorum
+func GetCephMonVersion(context *clusterd.Context) (*cephver.CephVersion, error) {
+	output, err := getCephMonVersionString(context)
+	if err != nil {
+		return nil, fmt.Errorf("failed to run ceph tell: %+v", err)
+	}
+	logger.Debug(output)
+
+	v, err := cephver.ExtractCephVersion(output)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract ceph version. %+v", err)
+	}
+
+	return v, nil
+}
+
+// GetCephVersions reports the Ceph version of each daemon in the cluster
+func GetCephVersions(context *clusterd.Context) (*CephDaemonsVersions, error) {
+	output, err := getCephMonVersionsString(context)
+	if err != nil {
+		return nil, fmt.Errorf("failed to run ceph tell: %+v", err)
+	}
+	logger.Debug(output)
+
+	var cephVersionsResult CephDaemonsVersions
+	err = json.Unmarshal([]byte(output), &cephVersionsResult)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve ceph versions results. %+v", err)
+	}
+
+	return &cephVersionsResult, nil
+}
+
+// EnableMessenger2 enable the messenger 2 protocol on Nautilus clusters
+func EnableMessenger2(context *clusterd.Context, namespace string) error {
+	args := []string{"mon", "enable-msgr2"}
+	_, err := client.ExecuteCephCommand(context, namespace, args)
+	if err != nil {
+		return fmt.Errorf("failed to enable msgr2 protocol: %+v", err)
+	}
+	logger.Infof("successfully enabled msgr2 protocol")
+
+	return nil
+}
+
+// OkToStopOrContinue determines wether it's ok to stop or continue an upgrade
+func OkToStopOrContinue(context *clusterd.Context, namespace, deployment, daemon, action string, cephVersion cephver.CephVersion) error {
+
+	// The ok-to-stop command for osd and mds landed on 14.2.1
+	// so we return nil if that Ceph version is not satisfied
+	if !cephVersion.IsAtLeast(cephver.CephVersion{Major: 14, Minor: 2, Extra: 1}) {
```

Comment:
The `ok-to-stop` command is implemented in the monitor (for mds and mon) and in the manager for osds. This check applies to the container with the `ceph` CLI command (IIRC). Since this is all in the context of an upgrade scenario, it might be important to ensure that the daemon implementing the `ok-to-stop` command (mon or mgr) has been upgraded to 14.2.1 before issuing the command.

Reply:
That's a good point, it's the operator that runs these commands. So it's expected to run an operator based on a new version of Ceph to perform the upgrade. This also means that these commands should not run if we are upgrading to Nautilus. Daemons must be Nautilus for this to work.

Thanks, I also realized I made a typo too. The OSD implementation was present before `14.2.1`, actually since Mimic.